### PR TITLE
feat: stableswap remove liquidity support

### PR DESCRIPTION
**Pull request by enthusiastmartin**
This PR adds support to multi asset liquidity removal.

Note that if all remaining shares are removed - pool is destroyed.

- [x] update weights
- [x] on change liquidity hook

### CODE REVIEW

File: pallets/stableswap/src/weights.rs [hunk 1]
```diff
@@ -181,6 +182,16 @@ impl WeightInfo for () {
 			.saturating_add(RocksDbWeight::get().reads(23_u64))
 			.saturating_add(RocksDbWeight::get().writes(7_u64))
 	}
+
+	fn remove_liquidity() -> Weight {
```
Comment (dmoka) [inline pallets/stableswap/src/weights.rs:186]: maybe it is still WIP, but this weight function has not been added to runtime `pallet_swableswap.rs` weights, i guess the benchmark generation on the ref machine is still left?!

File: pallets/stableswap/src/lib.rs [hunk 2]
```diff
@@ -878,6 +880,113 @@ pub mod pallet {
...
+				Pools::<T>::remove(pool_id);
+				let _ = AssetTradability::<T>::clear_prefix(pool_id, MAX_ASSETS_IN_POOL, None);
+				T::DustAccountHandler::remove_account(&Self::pool_account(pool_id))?;
+
+				Self::deposit_event(Event::PoolDestroyed { pool_id });
+
+				//TODO: consider adding hooks to clean up oracle!?
+			}
+
+			Self::deposit_event(Event::LiquidityRemoved {
```
Comment (dmoka) [inline pallets/stableswap/src/lib.rs:977]: please add a unit test check for depositing this event

File: pallets/stableswap/src/lib.rs [hunk 3]
```diff
@@ -878,6 +880,113 @@ pub mod pallet {
...
+			// it is tricky to update the hooks as it works with share prices of each asset, which is not possible to calculate with total issuance == 0.
+			// we should consider to introduce new hook to inform that this pool is being removed.
+			if share_amount != share_issuance {
+				Self::call_on_liquidity_change_hook(pool_id, &initial_reserves, share_issuance)?;
+			} else {
+				Pools::<T>::remove(pool_id);
+				let _ = AssetTradability::<T>::clear_prefix(pool_id, MAX_ASSETS_IN_POOL, None);
+				T::DustAccountHandler::remove_account(&Self::pool_account(pool_id))?;
+
+				Self::deposit_event(Event::PoolDestroyed { pool_id });
```
Comment (dmoka) [inline pallets/stableswap/src/lib.rs:972]: please add a unit test check for depositing this event

File: pallets/stableswap/src/lib.rs [hunk 4]
```diff
@@ -878,6 +880,113 @@ pub mod pallet {
...
+
+			// All done and updated. let's call the on_liquidity_changed hook.
+			// TODO: liquidity change hook does not work when we are removing all liquidity.
+			// it is tricky to update the hooks as it works with share prices of each asset, which is not possible to calculate with total issuance == 0.
+			// we should consider to introduce new hook to inform that this pool is being removed.
+			if share_amount != share_issuance {
+				Self::call_on_liquidity_change_hook(pool_id, &initial_reserves, share_issuance)?;
+			} else {
+				Pools::<T>::remove(pool_id);
+				let _ = AssetTradability::<T>::clear_prefix(pool_id, MAX_ASSETS_IN_POOL, None);
```
Comment (dmoka) [inline pallets/stableswap/src/lib.rs:969]: Why is this line needed? Anyway, it is not tested either, pls consider adding an assertion if possible.

File: pallets/stableswap/src/tests/remove_liquidity.rs [hunk 5]
```diff
@@ -992,3 +993,199 @@ fn removing_liquidity_with_exact_amount_should_apply_fee() {
...
+				pool_id,
+				shares,
+				BoundedVec::try_from(min_amounts).unwrap(),
+			));
+
+			assert_balance!(ALICE, asset_a, pool_a_balance);
+			assert_balance!(ALICE, asset_b, pool_b_balance);
+			assert_balance!(ALICE, asset_c, pool_c_balance);
+
+			// Ensure that has been removed
```
Comment (dmoka) [inline pallets/stableswap/src/tests/remove_liquidity.rs:1186]:
```suggestion
			// Ensure that pool has been removed
```

File: pallets/stableswap/src/lib.rs [hunk 6]
```diff
@@ -878,6 +880,113 @@ pub mod pallet {
...
+			let initial_reserves = pool
+				.reserves_with_decimals::<T>(&pool_account)
+				.ok_or(Error::<T>::UnknownDecimals)?;
+			let share_issuance = T::Currency::total_issuance(pool_id);
+
+			ensure!(
+				share_issuance == share_amount
+					|| share_issuance.saturating_sub(share_amount) >= T::MinPoolLiquidity::get(),
+				Error::<T>::InsufficientLiquidityRemaining
+			);
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/stableswap/src/lib.rs:916]: i dont think this is necessary. it is a simple check. same as the other one

File: pallets/stableswap/src/lib.rs [hunk 7]
```diff
@@ -878,6 +880,113 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		#[pallet::call_index(10)]
+		#[pallet::weight(<T as Config>::WeightInfo::remove_liquidity())]
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/stableswap/src/lib.rs:885]: no need.we benchmark for max assets in pool. which is 5 atm, so wont be big difference. 

can create an issue to rework but it would be needed for all other calls too

File: pallets/stableswap/src/weights.rs (hunk 1 above)
Comment (enthusiastmartin) [reply pallets/stableswap/src/weights.rs:186]: will be rebenchamrked

File: pallets/stableswap/src/lib.rs [hunk 8]
```diff
@@ -878,6 +880,113 @@ pub mod pallet {
...
+				ensure!(
+					Self::is_asset_allowed(pool_id, *asset_id, Tradability::REMOVE_LIQUIDITY),
+					Error::<T>::NotAllowed
+				);
+				let min_amount = min_amounts_out.remove(asset_id).ok_or(Error::<T>::IncorrectAssets)?;
+				let reserve = T::Currency::free_balance(*asset_id, &pool_account);
+
+				// Special case when withdrawing all remaining pool shares, so we can directly send all the remaining assets to the user.
+				let amount = if share_amount == share_issuance {
+					// no need to ensure the min amounts in this case
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/stableswap/src/lib.rs:941]: actually, i added the check as user dont need to be aware that all remaining liquidity is being removed.

File: pallets/stableswap/src/lib.rs (hunk 4 above)
Comment (enthusiastmartin) [reply pallets/stableswap/src/lib.rs:969]: we destroy the pool and clear all related storages when all liquidity from this pool is removed.

**Review by dmoka — APPROVED** [review]
